src/server/session.ts

```diff
@@ -294,71 +294,47 @@ namespace ts.server {
-            addToCombineOutputResult(outputs, project, filter(theseOutputs, output => !combineOutputResultContains(outputs, output, areEqual)));
-        });
-        return outputs;
-    }
-
-    function flattenCombineOutputResult<T>(outputs: CombineOutputResult<T>): readonly T[] {
-        return flatMap(outputs, ({ result }) => result);
-    }
-
-    function combineProjectOutputWhileOpeningReferencedProjects<T>(
-        projects: Projects,
-        defaultProject: Project,
-        action: (project: Project) => readonly T[],
-        getLocation: (t: T) => DocumentPosition,
-        resultsEqual: (a: T, b: T) => boolean,
-    ): CombineOutputResult<T> {
-        const outputs: CombineOutputResult<T> = [];
-        combineProjectOutputWorker(
-            projects,
-            defaultProject,
-            /*initialLocation*/ undefined,
-            (project, _, tryAddToTodo) => {
-                const theseOutputs = action(project);
-                addToCombineOutputResult(outputs, project, filter(theseOutputs, output => !combineOutputResultContains(outputs, output, resultsEqual) && !tryAddToTodo(project, getLocation(output))));
-            },
-        );
-        return outputs;
-    }
+    interface ProjectNavigateToItems {
+        project: Project;
+        navigateToItems: readonly NavigateToItem[];
+    };
 
-    function combineProjectOutputForRenameLocations(
+    function getRenameLocationsWorker(
         projects: Projects,
         defaultProject: Project,
-        initialLocation: DocumentPosition,
+        initialPosition: DocumentPosition,
         findInStrings: boolean,
         findInComments: boolean,
-        hostPreferences: UserPreferences
+        { providePrefixAndSuffixTextForRename }: UserPreferences
     ): readonly RenameLocation[] {
-        const outputs: RenameLocation[] = [];
-        combineProjectOutputWorker(
+        const perProjectResults = getPerProjectReferences(
             projects,
             defaultProject,
-            initialLocation,
-            (project, location, tryAddToTodo) => {
-                for (const output of project.getLanguageService().findRenameLocations(location.fileName, location.pos, findInStrings, findInComments, hostPreferences.providePrefixAndSuffixTextForRename) || emptyArray) {
-                    if (!contains(outputs, output, documentSpansEqual) && !tryAddToTodo(project, documentSpanLocation(output))) {
-                        outputs.push(output);
-                    }
-                }
-            },
+            initialPosition,
+            (project, position) => project.getLanguageService().findRenameLocations(position.fileName, position.pos, findInStrings, findInComments, providePrefixAndSuffixTextForRename),
+            (renameLocation, cb) => cb(documentSpanLocation(renameLocation)),
         );
 
-        return outputs;
+        // No filtering or dedup'ing is required if there's exactly one project

```

Comment:
Special-casing the single-project case is new and should avoid a bunch of linear-search de-duping.